File: array/0018_4_sum/0018_4_sum.py
import collections

class Solution(object):
    def fourSum(self, nums, target):
        nums, result, lookup = sorted(nums), [], collections.defaultdict(list)
        for i in xrange(0, len(nums) - 1):
            for j in xrange(i+1, len(nums)):
                lookup[nums[i]+nums[j]].append([i,j])

        for i in lookup.keys():
            if target - i in lookup:
                for x in lookup[i]:
                    for y in lookup[target-i]:
                        [a,b],[c,d] = x,y
                        if a is not c and a is not d and b is not c and b is not d:
                            quad = sorted([nums[a], nums[b], nums[c],nums[d]])
                            if quad not in result:
                                result.append(quad)
        return sorted(result)

nums = [1,0,-1,0,-2,2]
target = 0 

res = Solution().fourSum(nums, target)
print(res)

File: array/0018_4_sum/0018_4_sum.go
package main 

import (
	"fmt"
	"sort"
)

func fourSum(nums []int, target int) [][]int {
	res := [][]int{}
	sort.Ints(nums)

	for i:= 0; i < len(nums) - 3; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		for j:= i+1; j < len(nums) - 2 ; j++ {
			if j > i + 1 && nums[j] == nums[j-1] {
				continue
			}

			l, r := j + 1, len(nums) - 1
			for l < r {
				s := nums[i] + nums[j] + nums[l] + nums[r]
				switch {
				case s < target:
					l++
				case s > target:
					r--
				default:
					res = append(res, []int{nums[i], nums[j], nums[l], nums[r]})
					l, r = next(nums, l, r)
				}
			}
		}
	}

	return res
}

func next(nums []int, l,r int) (int, int) {
	for l < r {
		switch {
		case nums[l] == nums[l+1]:
			l++
		case nums[r] == nums[r-1]:
			r--
		default:
			l++
			r--
			return l,r
		}
	}

	return l, r
}

func main() {
	nums := []int{1,0,-1,0,-2,2}
	target := 0

	res := fourSum(nums, target)
	fmt.Println(res)
}